synapse/handlers/user_directory.py

```diff
@@ -217,30 +229,32 @@ async def _handle_deltas(self, deltas: List[Dict[str, Any]]) -> None:
                     else:
                         logger.debug("Server is still in room: %r", room_id)
 
-                is_support = await self.store.is_support_user(state_key)
-                if not is_support:
-                    if change is None:
-                        # Handle any profile changes
+                is_support_user = await self.store.is_support_user(state_key)
+                if is_support_user:
+                    continue
```

Comment:
```suggestion
                if is_support_user:
                    # support users should not be added to the directory
                    continue
```

just in case we wonder what this is for (since it won't be long before we all forget what support users are again..)

Reply:
They're users that really shouldn't be counted as users and shouldn't be exposed to real users on the server.

The common use case is support: EMS (presumably) might have these accounts on a customer's homeserver so that they can log in, check that things are working (do messages send? how long does it take? It's possible [matrixmon](https://github.com/matrix-org/matrixmon/) uses a support user on these servers for automatic testing of 'Can I send an event and receive it?'), etc

Reply:
Maybe the right thing to do (analogous to your suggestion elsewhere) is to separate the management of `user_directory` from the management of the `users_in_public_rooms` and `users_who_share_private_rooms`. The former can skip support users fairly safely I think?

Reply:
Without being able to fully imagine the difficulty of doing that cleanly, sounds sensible.

That said, `users_in_public_rooms` and `users_who_share_private_rooms` are *purely* for the user directory, so support users can be skipped from those as well.

I think, for your sanity, it should be sensible to just try and forget about support users as soon as you can. `continue` as you have it seems reasonable.

(do we even need this check here though? We should be preventing support users from being added to the directory already, and the plan is to not add local users to the directory from this background process, so thinking further, I think you can just forget about special casing them now?)

Reply:
> That said, users_in_public_rooms and users_who_share_private_rooms are purely for the user directory, so support users can be skipped from those as well.

Slightly anxious that this isn't obvious from the table names. But I'm struggling to think of a reason where you would actively want to count or search for these support users.

Yeah. Let's suppress them!

Reply:
Further to this: later on we skip adding users to the `users_who_share_private_room` and `users_in_public_room` tables. I think we can exclude these at the same time that we exclude support users. Do you agree?

Reply:
sounds reasonable